### ISSUE TITLE
Getting Users with Offset causes nothing to be returned.

### ISSUE THREAD

**Issue by mascon** [jhoerr/box-csharp-sdk-v2#3]
Dear Reader

I have stumbled across an issue with the GetUsers call.

In the code sample below you will see I loop around collecting 100 users at a time and displaying them on screen. 

Everything works find for the fist 10 calls but when the call for the offset value 1100 nothing is returned.

Not being a C# developer stepping through the code I did not notice anything obvious.  I have check the API from box and the information is returned without errors.

```
    Dim BoxClient As New BoxManager(BoxAccount.API_Key, BoxAccount.Auth_Token)

    Dim BoxUsers = BoxClient.GetUsers("", 1, 0)

    TotalBoxUsers = BoxUsers.TotalCount

    Do Until CurrentBoxUsersReturned >= TotalBoxUsers

        BoxUsers = BoxClient.GetUsers("", 1, BoxUserReturnedOffset)

        CurrentBoxUsersReturned = CurrentBoxUsersReturned + BoxUsers.Entries.Count

        CurrentUserIndex = 0

        Do Until CurrentUserIndex >= BoxUsers.Entries.Count - 1

            Console.Write(BoxUsers.Entries(CurrentUserIndex).Login & vbNewLine)

            CurrentUserIndex = CurrentUserIndex + 1

        Loop

        BoxUserReturnedOffset = BoxUserReturnedOffset + 100

        BoxUsers = Nothing

    Loop 
```

Any assistance in getting this solved within the SDK would be helpful.

Thanks in advance

Simon

**Comment by jhoerr** [comment]
Hi Simon,

Thanks for the heads up!  I'll look into this today.

John

**Comment by jhoerr** [comment]
Simon,

When you say that you 'have checked the API from box and the information is returned without errors', do I understand correctly that you've attempted the same request against the RESTful API and it succeeded?

**Comment by mascon** [comment]
Hi John

Yep I have made the calls from a restful client and mirroring exactly what I thought the sdk was doing. All went perfect. 

Another thing I believe I noticed when using the SDK was if I up the limit to say 500 only 100 was returned. Max from the API perspective for the limit is 1000. 

Are you able to recreate the issue?

Cheers

Simon
Sent from my BlackBerry® wireless device

-----Original Message-----

**Comment by jhoerr** [comment]
I have not been able to reproduce it yet.  Our organization (luckily) has an Enterprise account with lots of users, so I should be able to.  But first I need to track down a particular colleague that has admin privileges.  In the meantime I'm reviewing the code and comparing it the the documentation.  Nothing is jumping out at me.

**Comment by mascon** [comment]
Hi John

When I stepped through it from what I could see was there was an error from the API complaining about incorrect attribute I seem to remember. It was on the line of code that returns the query and casts it into the user object before returning to the caller. 

Hope that helps

Simon

Sent from my BlackBerry® wireless device

-----Original Message-----

**Comment by jhoerr** [comment]
Ok, thanks for that tip.  I've got a test set up for the failing condition.  I'll run it as soon as I can get an admin token and hopefully that'll get me pointed in the right direction.

**Comment by jhoerr** [comment]
Simon, I'm having trouble reproducing this.  We have many thousand users in our enterprise, which should provide a sufficient test bed.  The only case that I've come across in which the client returns 0 users is when the specified offset exceeds the number of users in the enterprise.

For sake of conversation, say we have 2050 users in our enterprise.

Client.GetUsers() -> returns users 1-100 (box returns 100 users by default)
Client.GetUsers(null, 0, 500) -> returns users 501-600 (100 users returned by default even though 0 requested)
Client.GetUsers(null, 1, 500) -> returns user 501
Client.GetUsers(null, 100, 0) -> returns users 1-100
Client.GetUsers(null, 1000, 0) -> returns users 1-1000
Client.GetUsers(null, 1500, 0) -> returns users 1-1000 (a max of 1000 users is returned for a single call)
Client.GetUsers(null, 100, 1000) -> returns users 1001-1100.
Client.GetUsers(null, 1000, 1000) -> returns users 1001-2000.
Client.GetUsers(null, 1000, 2000) -> returns users 2001-2050
Client.GetUsers(null, 1000, 3000) -> returns no users, because the offset exceeds the total number of users.

> Another thing I believe I noticed when using the SDK was if I up the limit to say 500 only 100 was returned. Max from the API perspective for the limit is 1000.

Can you provide a snippet of code that causes this behavior?

**Comment by mascon** [comment]
Hi John

I have attached a screen shot of the error I see deep within the C# code.  
Not sure if this gives you any clues.

I can offer a screen share with you if that helps. (Will be a little slow as 
I am home)

Regards

Simon

---

No trees were harmed in the sending of this
email, but a large number of electrons were
terribly inconvenienced.

-----Original Message-----

**Comment by jhoerr** [comment]
I don't think the attachment made it through.  Can you email me the
screenshot directly a <my github username> at gmail dot com?

On Fri, Dec 14, 2012 at 3:41 PM, mascon notifications@github.com wrote:

> Hi John
> 
> I have attached a screen shot of the error I see deep within the C# code.
> Not sure if this gives you any clues.
> 
> I can offer a screen share with you if that helps. (Will be a little slow
> as
> I am home)
> 
> Regards
> 
> Simon
> 
> ---
> 
> No trees were harmed in the sending of this
> email, but a large number of electrons were
> terribly inconvenienced.
> 
> -----Original Message-----
> From: John Hoerr notifications@github.com
> To: jhoerr/box-csharp-sdk-v2 box-csharp-sdk-v2@noreply.github.com
> Cc: mascon imom@andimgone.com
> Date: Fri, 14 Dec 2012 12:16:23 -0800
> Subject: Re: [box-csharp-sdk-v2] Getting Users with Offset causes nothing
> to
> be returned. (#3)
> 
> Simon, I'm having trouble reproducing this. We have many thousand users in
> our enterprise, which should provide a sufficient test bed. The only case
> that I've come across in which the client returns 0 users is when the
> specified offset exceeds the number of users in the enterprise.
> 
> For sake of conversation, say we have 2050 users in our enterprise.
> 
> Client.GetUsers() -> returns users 1-100 (box returns 100 users by default)
> Client.GetUsers(null, 0, 500) -> returns users 501-600 (100 users returned
> by default even though 0 requested)
> Client.GetUsers(null, 1, 500) -> returns user 501
> Client.GetUsers(null, 100, 0) -> returns users 1-100
> Client.GetUsers(null, 1000, 0) -> returns users 1-1000
> Client.GetUsers(null, 1500, 0) -> returns users 1-1000 (a max of 1000
> users
> is returned for a single call)
> Client.GetUsers(null, 100, 1000) -> returns users 1001-1100.
> Client.GetUsers(null, 1000, 1000) -> returns users 1001-2000.
> Client.GetUsers(null, 1000, 2000) -> returns users 2001-2050
> Client.GetUsers(null, 1000, 3000) -> returns no users, because the offset
> exceeds the total number of users.
> 
> Another thing I believe I noticed when using the SDK was if I up the limit
> to say 500 only 100 was returned. Max from the API perspective for the
> limit
> is 1000.
> 
> Can you provide a snippet of code that causes this behavior?
> 
> —
> Reply to this email directly or view it on GitHub.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/jhoerr/box-csharp-sdk-v2/issues/3#issuecomment-11391485.

## 

John Hoerr
jhoerr@gmail.com
503.360.6581

**Comment by jhoerr** [comment]
Thanks!  Based on that I think you're doing everything properly.  The request is generating a valid response (you can see the JSON data coming back in the Content property.) This looks like it could be a JSON deserialization issue.  Do you use tracking codes, perchance?

**Comment by mascon** [comment]
## Yep two of them

No trees were harmed in the sending of this
email, but a large number of electrons were
terribly inconvenienced.

-----Original Message-----

**Comment by jhoerr** [comment]
Ok, that might have something to do with it.  I'll work on this a bit more and see what I can figure out.

**Comment by mascon** [comment]
If you need more information just shout.  I see you have added support for 
the updateuser to the SDK today.

I have another question as I can not see how to handle it.

In Enterprise land we have two hidden API V1 calls

create_admin_access_tokens and delete_admin_access_tokens

These have yet to be ported to V2 API but they allow admins to "Become" a 
user to act on their behalf during API calls.

I did not see a way to change the Auth_Token once the manager object was 
created.  Is this possible?

**Comment by jhoerr** [comment]
Yeah, I got the User update stuff in without too much fuss.  The Boxers don't make too much noise when they add new features like that, so thanks again for letting me know about it.

> I did not see a way to change the Auth_Token once the manager object was created.

It's not possible to modify the auth token via the (public) V2 API.  (And I'm not being cagey when I say that -- I simply have no access to the private APIs.)  What you might do is use the private V1 API to modify the token and then with that modified token create a new instance on the V2 SDK.

**Comment by jhoerr** [comment]
Fixed in 2d04fdaee9
